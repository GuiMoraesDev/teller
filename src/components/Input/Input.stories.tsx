import React from 'react'

import { ComponentMeta, ComponentStory } from '@storybook/react'

import Input from '.'
import { Plus } from 'phosphor-react'

const Storie = {
  title: 'Components/Input',
  component: Input
} as ComponentMeta<typeof Input>

export default Storie

const Template: ComponentStory<typeof Input> = (args) => <Input {...args} />

export const Default = Template.bind({})

Default.args = {
  label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm'
}

export const DefaultRequired = Template.bind({})

DefaultRequired.args = {
  label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm',
  isRequired: true
}

export const DefaultNoLabel = Template.bind({})

DefaultNoLabel.args = {
  dimension: 'sm',
  rounded: 'sm'
}

export const DefaultWithError = Template.bind({})

DefaultWithError.args = {
  label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm',
  error: 'required'
}

export const Password = Template.bind({})

Password.args = {
  label: 'Random Input',
  type: 'password',
  dimension: 'sm',
  rounded: 'sm'
}


export const DefaultIconLeft = Template.bind({})

DefaultIconLeft.args = {
	label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm',
	PlaceholderIconLeft: <Plus />,
}

export const DefaultIconRight = Template.bind({})

DefaultIconRight.args = {
  label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm',
	PlaceholderIconRight: <Plus />,
}

export const DefaultIconLeftAndRight = Template.bind({})

DefaultIconLeftAndRight.args = {
  label: 'Random Input',
  dimension: 'sm',
  rounded: 'sm',
	PlaceholderIconLeft: <Plus />,
	PlaceholderIconRight: <Plus />,
}
