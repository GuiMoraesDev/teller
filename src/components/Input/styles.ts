import styled, { css } from "styled-components";

import { InputDefaultPropsThatMakeStyles } from ".";

export const Container = styled.div`
  position: relative;

  width: ${({ theme }) => theme.sizes.partition["x1/1"]};
`;

export const Label = styled.label``;

export const InputContainer = styled.label<InputDefaultPropsThatMakeStyles>`
  position: relative;

  display: flex;

  align-items: center;

  width: auto;

  gap: ${({ theme }) => theme.sizes.common.x1};

  color: ${({ theme }) => theme.themeColors.text};

  background-color: ${({ theme }) => theme.themeColors.canvasDark};

  border: ${({ theme }) => theme.borders.solid};

  border-color: ${({ theme, error }) => error && theme.colors.alert["500"]};

  transition: box-shadow ${({ theme }) => theme.transition.fast},
    border-color ${({ theme }) => theme.transition.fast};

  &:hover,
  &:focus-within {
    border-color: ${({ theme, error }) =>
      error ? theme.colors.alert[300] : theme.themeColors.borderInverted};
    box-shadow: ${({ theme }) => theme.shadows.short};
  }

  ${({ theme, rounded }) => {
    return css`
      border-radius: ${theme.rounded[rounded || "none"]};
    `;
  }}

  ${({ dimension, theme }) => {
    if (dimension === "xs")
      return css`
        ${theme.typography.variants.body3};

        height: ${theme.sizes.common.x9};
      `;

    if (dimension === "sm")
      return css`
        ${theme.typography.variants.body3};

        height: ${theme.sizes.common.x10};
      `;

    if (dimension === "md")
      return css`
        ${theme.typography.variants.body1};

        height: ${theme.sizes.common.x11};
      `;

    if (dimension === "lg")
      return css`
        ${theme.typography.variants.body1};

        height: ${theme.sizes.common.x12};
      `;

    if (dimension === "xl")
      return css`
        ${theme.typography.variants.body1};

        height: ${theme.sizes.common.x14};
      `;
  }}

  ${({ theme, isDisabled }) => {
    if (isDisabled)
      return css`
        background-color: ${theme.colors.neutrals[200]};
        color: ${theme.colors.neutrals[500]};
        box-shadow: 0 0 0 1px ${theme.colors.neutrals[200]};

        &:hover {
          background-color: ${theme.colors.neutrals[200]};
          color: ${theme.colors.neutrals[500]};
        }
      `;
  }}
`;

export const InputComponent = styled.input`
  outline: none;

  width: 100%;
  height: 100%;

  padding: ${({ theme }) =>
    `${theme.sizes.common.x0} ${theme.sizes.common.x4}`};

  color: inherit;
  background-color: inherit;

  border: none;
  border-radius: inherit;
`;

export const ActionButton = styled.button`
  position: absolute;
  top: ${({ theme }) => theme.sizes.partition["x1/2"]};
  right: ${({ theme }) => theme.sizes.common.x2};

  display: flex;

  align-items: center;
  justify-content: center;

  margin: ${({ theme }) => theme.sizes.common.x0};
  padding: ${({ theme }) => theme.sizes.common.x0};

  background-color: transparent;

  border: none;

  transform: translateY(-${({ theme }) => theme.sizes.partition["x1/2"]});
  transition: transform ${({ theme }) => theme.transition.fast} ease-in-out;

  > .icon {
    width: ${({ theme }) => theme.sizes.common["x4.5"]};
    height: ${({ theme }) => theme.sizes.common["x4.5"]};

    fill: ${({ theme }) => theme.colors.neutrals["500"]};

    opacity: ${({ theme }) => theme.opacity.mid};

    transition: opacity ${({ theme }) => theme.transition.fast};
  }

  &:hover > .icon {
    opacity: 1;

    fill: ${({ theme }) => theme.colors.neutrals["900"]};
  }
`;

interface ErrorMessageProps {
  error?: string;
}

export const ErrorMessage = styled.span<ErrorMessageProps>`
  position: absolute;
  bottom: ${({ theme }) => theme.sizes.common["x0.5"]};
  right: ${({ theme }) => theme.sizes.common.x0};
  left: ${({ theme }) => theme.sizes.common.x0};

  display: flex;

  align-items: center;

  width: ${({ theme }) => theme.sizes.partition["x1/1"]};

  ${({ theme }) => theme.typography.variants.body3};

  margin: auto;
  padding: ${({ theme }) =>
    `${theme.sizes.common["x1.5"]} ${theme.sizes.common.x3}`};

  color: ${({ theme }) => theme.colors.neutrals["000"]};
  background-color: ${({ theme }) => theme.colors.alert["300"]};

  border-radius: 0;
  border-bottom-left-radius: ${({ theme }) => theme.rounded.sm};
  border-bottom-right-radius: ${({ theme }) => theme.rounded.sm};

  opacity: ${({ error }) => (error ? 1 : 0)};
  transform: translateY(100%);
  transition: opacity 0.5s ease-in-out;

  z-index: 2;

  svg {
    margin-right: 0.5rem;
  }
`;
