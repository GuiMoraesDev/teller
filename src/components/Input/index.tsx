import React, { InputHTMLAttributes } from "react";

import { Warning } from "phosphor-react";
import { DefaultTheme } from "styled-components";

import Text from "components/Text";

import * as Styles from "./styles";

export interface InputDefaultPropsThatMakeStyles {
  dimension?: "xs" | "sm" | "md" | "lg" | "xl";
  rounded?: keyof DefaultTheme["rounded"];
  isDisabled?: boolean;
  error?: string;
}

interface InputDefaultProps
  extends InputDefaultPropsThatMakeStyles,
    InputHTMLAttributes<HTMLInputElement> {
  id: string;
  placeholder?: string;
  className?: string;
  label?: string;
  description?: string;
  type?: string;
  required?: boolean;
  optional?: boolean;
  handleOnChange?(value: string): void;
}

const Input: React.ForwardRefRenderFunction<
  HTMLInputElement,
  InputDefaultProps
> = (
  {
    id,
    error,
    label,
    description,
    type,
    placeholder,
    required = false,
    optional = true,
    handleOnChange,
    dimension = "md",
    rounded = "sm",
    className,
    ...props
  },
  ref
) => {
  const handleChangeValue = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      props.onChange?.(event);

      const { value } = event.target;

      handleOnChange?.(value);
    },
    [handleOnChange, props]
  );

  return (
    <Styles.Container className={className}>
      {label && (
        <Styles.Label htmlFor={id}>
          <Text label={label} dimension="body1" />
        </Styles.Label>
      )}

      <Styles.InputContainer dimension={dimension} rounded={rounded}>
        <Styles.InputComponent
          placeholder={placeholder || "Type here..."}
          ref={ref}
          id={id}
          disabled={props.isDisabled}
          {...props}
          onChange={handleChangeValue}
        />
      </Styles.InputContainer>

      <Styles.ErrorMessage error={error}>
        <Warning />
        {error}
      </Styles.ErrorMessage>
    </Styles.Container>
  );
};

export default React.forwardRef(Input);
