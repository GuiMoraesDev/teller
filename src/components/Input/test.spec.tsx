import React from "react";

import { render } from "tests/test.utils";

import Input from "./index";

describe("Input component", () => {
  it("should render correctly", () => {
    const { container } = render(<Input id="test-input" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render visible", () => {
    const { container } = render(<Input id="test-input" />);

    expect(container.firstChild).toBeVisible();
  });
});
