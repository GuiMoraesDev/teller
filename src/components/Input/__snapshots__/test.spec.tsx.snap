// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render correctly 1`] = `
.c0 {
  position: relative;
  width: 100%;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  gap: 0.4rem;
  color: #181917;
  background-color: #E4E5E3;
  border: 0.1rem solid #C9CAC8;
  -webkit-transition: box-shadow 0.2s, border-color 0.2s;
  transition: box-shadow 0.2s, border-color 0.2s;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 2.4rem;
  height: 4.4rem;
}

.c1:hover,
.c1:focus-within {
  border-color: #797B75;
  box-shadow: 1px 0px 3px 0px #E4E5E3;
}

.c2 {
  outline: none;
  width: 100%;
  height: 100%;
  padding: 0 1.6rem;
  color: inherit;
  background-color: inherit;
  border: none;
  border-radius: inherit;
}

.c3 {
  position: absolute;
  bottom: 0.2rem;
  right: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  font-size: 1.2rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 1.6rem;
  margin: auto;
  padding: 0.6rem 1.2rem;
  color: #FFFFFF;
  background-color: #FFE3A3;
  border-radius: 0;
  border-bottom-left-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
  opacity: 0;
  -webkit-transform: translateY(100%);
  -ms-transform: translateY(100%);
  transform: translateY(100%);
  -webkit-transition: opacity 0.5s ease-in-out;
  transition: opacity 0.5s ease-in-out;
  z-index: 2;
}

.c3 svg {
  margin-right: 0.5rem;
}

<div
  class="c0"
>
  <label
    class="c1"
  >
    <input
      class="c2"
      id="test-input"
      placeholder="Type here..."
    />
  </label>
  <span
    class="c3"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        fill="none"
        height="256"
        width="256"
      />
      <line
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
        x1="128"
        x2="128"
        y1="104"
        y2="144"
      />
      <path
        d="M114.2,40l-88,152A16,16,0,0,0,40,216H216a16,16,0,0,0,13.8-24l-88-152A15.9,15.9,0,0,0,114.2,40Z"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="16"
      />
      <circle
        cx="128"
        cy="180"
        r="12"
      />
    </svg>
  </span>
</div>
`;
